src/context/blocked.rs

```diff
@@ -1,18 +1,177 @@
+        Self {
+            inner,
+            reply_callback: None,
+            free_priv_callback: None,
+            private_data: None,
+        }
+    }
 
-impl Drop for BlockedClient {
+    #[allow(dead_code)]
+    pub(crate) fn with_auth_callback(
+        inner: *mut raw::RedisModuleBlockedClient,
+        auth_reply_callback: fn(
+            &Context,
+            ValkeyString,
+            ValkeyString,
+            T,
+        ) -> Result<c_int, ValkeyError>,
+        free_priv_callback: Option<FreePrivDataCallback<T>>,
+    ) -> Self {
+        Self {
+            inner,
+            reply_callback: Some(ReplyCallback::Auth(auth_reply_callback)),
+            free_priv_callback,
+            private_data: None,
+        }
+    }
+
+    /// Sets private data for the blocked client.
+    ///
+    /// # Warning
+    /// Currently, this method should only be used with `block_client_on_auth`.
+    /// Using this with general blocking operations (via `block_client`) may cause memory leaks
+    /// as there is no way to free the private data in the general blocking case yet.
+    /// Support for general blocking with proper memory management will be added in a future (TODO).
+    ///
+    /// # Arguments
+    /// * `data` - The private data to store
+    pub fn set_blocked_private_data(&mut self, data: T) {
+        self.private_data = Some(Box::new(data));
+    }
+
+    /// Aborts the blocked client operation
+    ///
+    /// # Returns
+    /// * `Ok(())` - If the blocked client was successfully aborted
+    /// * `Err(ValkeyError)` - If the abort operation failed
+    pub fn abort(mut self) -> Result<(), ValkeyError> {
+        unsafe {
+            // Clear references to private_data and callbacks
+            self.private_data = None;
+            self.reply_callback = None;
+            self.free_priv_callback = None;
+
+            if raw::RedisModule_AbortBlock.unwrap()(self.inner) == raw::REDISMODULE_OK as c_int {
+                // Prevent the normal Drop from running
+                self.inner = std::ptr::null_mut();
+                Ok(())
+            } else {
+                Err(ValkeyError::Str("Failed to abort blocked client"))
+            }
+        }
+    }
+}
+
+impl<T> Drop for BlockedClient<T> {
     fn drop(&mut self) {
-        unsafe { raw::RedisModule_UnblockClient.unwrap()(self.inner, ptr::null_mut()) };
+        if !self.inner.is_null() {
+            let callback_data_ptr = self
+                .private_data
+                .take()
+                .map(|private_data| {
+                    let callbacks_and_data = (
+                        self.reply_callback.take(),
+                        self.free_priv_callback.take(),
+                        private_data,
+                    );
+                    Box::into_raw(Box::new(callbacks_and_data)) as *mut c_void
+                })
+                .unwrap_or_else(std::ptr::null_mut);
```

Comment:
Do we need both `take` and `map`? Doesn't `map` itself consume the `Option` (so we could use `map` instead of both operations)?

If we really do need both, would suggest splitting this into two lines


Also, I don't think we need `unwrap_or_else` since we could use `map_or_else` for this entire transformation